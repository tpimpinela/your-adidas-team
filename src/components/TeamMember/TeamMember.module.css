.team-member {
  position: relative;
  transition: 0.2s transform, filter 0.2s, opacity 0.2s;

  &.team-member--clicable {
    cursor: pointer;

    &:hover {
      transform: scale(1.02);
    }
  }

  &.team-member--disabled {
    pointer-events: none;
    cursor: not-allowed;
    filter: grayscale(1);
    opacity: 0.9;
  }

  & .included-mark {
    position: absolute;
    top: 0;
    right: 0;
    transform: translateX(20%) translateY(-20%);
    font-size: 1.5rem;
    padding: 0.2rem;
    border-radius: 2px;
    background-color: var(--accent-color);
    color: white;
  }
  & .team-member__image {
    border-radius: 15px 15px 0 0;
    width: 100%;
    height: 100%;
    object-fit: cover;
    padding-bottom: 1.5rem;
  }

  & .team-member__info {
    position: absolute;
    bottom: 0;
    left: 0;
    width: 100%;
    box-sizing: border-box;
    display: flex;
    flex-direction: column;

    & .name {
      padding: 0.5rem;
      padding-left: 1.2rem;
      background-color: var(--accent-color);
      border-top-left-radius: 15px;
    }

    & .position {
      background-color: var(--second-background-color);
      font-size: 0.8rem;
      padding: 0.2rem;
      text-align: right;
      text-transform: uppercase;
    }
  }
}
