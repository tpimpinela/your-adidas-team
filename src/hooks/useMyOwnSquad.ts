import { useCallback, useContext } from "react";
import { MyOwnSquadContext } from "../contexts/MyOwnSquad.context";

const useMyOwnSquad = () => {
  const myOwnSquad = useContext(MyOwnSquadContext);

  const isMemberOnMySquad = useCallback(
    (teamMemberId: number) => {
      return Boolean(myOwnSquad.find(({ id }) => id === teamMemberId));
    },
    [myOwnSquad]
  );

  return {
    squad: myOwnSquad,
    isMemberOnMySquad,
  };
};

export default useMyOwnSquad;
