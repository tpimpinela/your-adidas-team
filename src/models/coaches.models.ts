import { FootballAPIResponse } from "./footballApiResponse.model";
import { TeamMapped } from "./teams.models";

export interface Coach {
  id: number;
  name: string;
  firstname: string;
  lastname: string;
  age: number;
  birth: Birth;
  nationality: string;
  height: null | string;
  weight: null | string;
  photo: string;
  team: TeamMapped;
  career: Career[];
}

export type CoachMapped = Pick<
  Coach,
  "id" | "name" | "age" | "photo" | "firstname" | "lastname"
>;

interface Birth {
  date: Date;
  place: null | string;
  country: string;
}

interface Career {
  team: TeamMapped;
  start: Date;
  end: Date | null;
}

export type FootbalCoachesApiResponse = FootballAPIResponse<Coach>;
