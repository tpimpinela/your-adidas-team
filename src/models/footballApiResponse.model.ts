export interface FootballAPIResponse<T> {
  get: string;
  parameters: Parameters;
  errors: Error[];
  results: number;
  paging: Paging;
  response: T[];
}

interface Paging {
  current: number;
  total: number;
}

interface Parameters {
  team: string;
}

interface Error {
  time: string;
  bug: string;
  report: string;
}
