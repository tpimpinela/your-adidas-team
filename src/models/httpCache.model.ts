import { FootballAPIResponse } from "./footballApiResponse.model";

export interface HttpCache {
  [fullEndpoint: string]: FootballAPIResponse<any> & { validUntil: Date };
}
