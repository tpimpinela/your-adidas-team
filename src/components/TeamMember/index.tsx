import { useMemo } from "react";
import useMyOwnSquad from "../../hooks/useMyOwnSquad";
import { CoachMapped } from "../../models/coaches.models";
import { PlayerMapped } from "../../models/squads.models";
import { TeamMember } from "../../models/teamMember.model";
import styles from "./TeamMember.module.css";

interface Props {
  teamMember: PlayerMapped | CoachMapped;
  onTeamMemberClicked?: (teamMember: TeamMember, isOnTheSquad: boolean) => void;
  disabled?: boolean;
}

const TeamMember = ({ teamMember, onTeamMemberClicked, disabled }: Props) => {
  const { isMemberOnMySquad: isMemberOnMySquadFn } = useMyOwnSquad();

  const isMemberOnMySquad = useMemo(
    () => isMemberOnMySquadFn(teamMember.id),
    [teamMember.id, isMemberOnMySquadFn]
  );

  return (
    <div
      onClick={() => onTeamMemberClicked?.(teamMember, isMemberOnMySquad)}
      className={`${styles["team-member"]} ${
        (!disabled && onTeamMemberClicked) || isMemberOnMySquad
          ? `${styles["team-member--clicable"]}`
          : `${styles["team-member--disabled"]}`
      }`}
      data-testid="team-member"
    >
      <img
        className={styles["team-member__image"]}
        src={teamMember.photo}
        alt={`Photo of ${teamMember.name}`}
      />
      <div className={styles["team-member__info"]}>
        <span className={styles.name}>{teamMember.name}</span>
        <span className={styles.position}>{teamMember.position}</span>
      </div>
      {isMemberOnMySquad && <span className={styles["included-mark"]}>âœ“</span>}
    </div>
  );
};

export default TeamMember;
