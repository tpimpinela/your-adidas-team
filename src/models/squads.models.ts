import { FootballAPIResponse } from "./footballApiResponse.model";
import { TeamMapped } from "./teams.models";

export interface Squad {
  team: TeamMapped;
  players: Player[];
}

export interface Player {
  id: number;
  name: string;
  age: number;
  number: number;
  position: PlayerPosition;
  photo: string;
}

export type PlayerMapped = Pick<
  Player,
  "id" | "name" | "number" | "position" | "photo"
>;

export enum PlayerPosition {
  Attacker = "Attacker",
  Defender = "Defender",
  Goalkeeper = "Goalkeeper",
  Midfielder = "Midfielder",
}

export type FootbalSquadsApiResponse = FootballAPIResponse<Squad>;
