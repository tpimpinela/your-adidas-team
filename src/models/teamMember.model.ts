import { CoachMapped } from "./coaches.models";
import { PlayerMapped } from "./squads.models";

export type TeamMember = PlayerMapped | CoachMapped;
