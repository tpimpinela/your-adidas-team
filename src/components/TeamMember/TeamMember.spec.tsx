import { cleanup, render, RenderResult } from "@testing-library/react";
import { beforeEach, describe, expect, it, Mock, Mocked, vi } from "vitest";
import TeamMember from ".";
import coachMock from "../../mocks/coach.mock";
import squadMock from "../../mocks/squads.mock";
import userEvent from "@testing-library/user-event";
import { PlayerMapped, PlayerPosition } from "../../models/squads.models";
import useMyOwnSquad from "../../hooks/useMyOwnSquad";

vi.mock("../../hooks/useMyOwnSquad", () => ({
  default: vi.fn(),
}));

describe("<TeamMember />", () => {
  let renderResult: RenderResult;
  const playerMock = squadMock.players[0] as unknown as PlayerMapped;
  let onTeamMemberClickedMock: Mocked<any>;
  const useMyOwnSquadMock = useMyOwnSquad as Mock<any, any>;
  const isMemberOnMySquadMock = vi.fn();

  beforeEach(() => {
    useMyOwnSquadMock.mockReturnValue({
      isMemberOnMySquad: isMemberOnMySquadMock,
    });
    isMemberOnMySquadMock.mockReturnValue(false);
    onTeamMemberClickedMock = vi.fn();
    renderResult = render(
      <TeamMember
        teamMember={playerMock}
        onTeamMemberClicked={onTeamMemberClickedMock}
      />
    );
  });

  it("should show teamMember name", () => {
    expect(renderResult.getByText(playerMock.name)).toBeTruthy();
  });

  it("should show teamMember image", () => {
    expect((renderResult.getByRole("img") as HTMLImageElement).src).toBe(
      playerMock.photo
    );
  });

  describe("if teamMember is a player", () => {
    beforeEach(() => {
      cleanup();
      renderResult = render(<TeamMember teamMember={playerMock} />);
    });

    it("should match the snapshot", () => {
      expect(renderResult.container).toMatchSnapshot();
    });

    it("should show player position", () => {
      expect(renderResult.getByText(playerMock.position)).toBeTruthy();
    });
  });

  describe("if teamMember is a coach", () => {
    beforeEach(() => {
      cleanup();
      renderResult = render(
        <TeamMember
          teamMember={{
            ...coachMock,
            teamId: 9,
            position: PlayerPosition.Coach,
          }}
        />
      );
    });

    it("should match the snapshot", () => {
      expect(renderResult.container).toMatchSnapshot();
    });

    it("shoud show coach literal", () => {
      expect(renderResult.getByText("Coach")).toBeTruthy();
    });
  });

  it("should call onTeamMemberClicked prop function if the member is clicked", async () => {
    await userEvent.click(renderResult.getByTestId("team-member"));
    expect(onTeamMemberClickedMock).toHaveBeenNthCalledWith(
      1,
      playerMock,
      false
    );
  });

  describe("marks", () => {
    it("should show delete mark", () => {
      renderResult.rerender(
        <TeamMember teamMember={playerMock} showDeleteMark />
      );
      expect(renderResult.getByTestId("delete-mark")).toBeTruthy();
    });

    it("should not show the delete mark if prop is falsy", () => {
      renderResult.rerender(
        <TeamMember teamMember={playerMock} showDeleteMark={false} />
      );
      expect(renderResult.queryByTestId("delete-mark")).toBeFalsy();
    });
  });
});
