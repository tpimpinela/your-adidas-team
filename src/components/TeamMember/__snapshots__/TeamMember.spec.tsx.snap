// Vitest Snapshot v1

exports[`<TeamMember /> > if teamMember is a coach > should match the snapshot 1`] = `
<div>
  <div
    class="_team-member_30838e _team-member--disabled_30838e"
    data-testid="team-member"
  >
    <img
      alt="Photo of Luis Enrique"
      class="_team-member__image_30838e"
      src="https://media-3.api-sports.io/football/coachs/193.png"
    />
    <div
      class="_team-member__info_30838e"
    >
      <span
        class="_name_30838e"
      >
        Luis Enrique
      </span>
      <span
        class="_position_30838e"
      >
        Coach
      </span>
    </div>
  </div>
</div>
`;

exports[`<TeamMember /> > if teamMember is a player > should match the snapshot 1`] = `
<div>
  <div
    class="_team-member_30838e _team-member--disabled_30838e"
    data-testid="team-member"
  >
    <img
      alt="Photo of Robert Lynch Sánchez"
      class="_team-member__image_30838e"
      src="https://media.api-sports.io/football/players/18959.png"
    />
    <div
      class="_team-member__info_30838e"
    >
      <span
        class="_name_30838e"
      >
        Robert Lynch Sánchez
      </span>
      <span
        class="_position_30838e"
      >
        Goalkeeper
      </span>
    </div>
  </div>
</div>
`;
